Multi Version Bridge API

Floweynt opened the pull request:
Provide an internal API over Minecraft to remove duplicated code across minecraft versions.

This is implemented via a set of interfaces, which are injected via loom interface injection (at compile time) and Mixins (at runtime). A stub class will be exposed to `common` as API, which provides static methods. These methods are overwritten via Mixin at runtime.

Goals:
- Abstract most version independent logic to `common`
- Reduce maintenance burden for new & old features

Non-goals:
- Provide a stable external interface to third-party modules
- Implement "multi-version" jars
- Have "perfect" abstraction - some leaky abstraction is allowed, but these must be for specific edgecases only, e.g. an OpenGL call that's only needed in 1.8.9 

Checklist:
- [x] Implement Bridge API for `1.8.9` (reference implementation for ports)
- [x] Implement Bridge API for `1.16`
- [x] Implement Bridge API for `1.20`
- [x] Implement Bridge API for `1.21`
- [x] Implement Bridge API for `1.21.4`
- [x] Port most HUD modules
- [ ] ~Port other modules~

During the porting process, `0` or some other suffix may be appended to avoid name collisions. These will be removed once colliding packages & classes have been abstracted away.

Commit:
feat: initial bridge API implementation (1.8.9, some hud modules port)

moehreag commented on 1.20/src/main/java/io/github/axolotlclient/modules/hud/gui/hud/simple/CustomHudEntry.java:
```diff
@@ -41,8 +41,8 @@ public class CustomHudEntry extends SimpleTextHudEntry {
 		removeEntry = new GenericOption("custom_hud.remove", "custom_hud.remove.label", () -> {
-			HudManager.getInstance().removeEntry(this.id);
-			HudManager.getInstance().saveCustomEntries();
+			//HudManager.getInstance().removeEntry(this.id);
```
It would probably make sense to add `TODO` to this to not forget to uncomment/fix this

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/bridge/impl/AxoRenderContextImpl.java:
```diff
@@ -0,0 +1,104 @@
+import net.minecraft.resource.Identifier;
+import org.jetbrains.annotations.Nullable;
+
+public class AxoRenderContextImpl implements AxoRenderContext {
```
On 1.8.9, this should ideally be an implementation akin to `GuiGraphics` on 1.20.1/1.21. On those newer versions, it should be a wrapper implementation. Another reason to prefer newer methods as a base are differences in parameter types and method usages, f.e. in regards to drawing textures (`blitSprite` methods) due to render types as well as text rendering (which isn't directly being delegated to the `TextRenderer`/`Font`). As such, a stateful matrix implementation might be necessary for 1.8.9 as well.

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/bridge/impl/AxoRenderContextImpl.java:
```diff
@@ -0,0 +1,104 @@
+	@Override
+	public void glDisableTexture() {
+		GlStateManager.disableTexture();
+	}
```
These should ideally be an implementation detail for 1.8.9 and used when drawing a texture, they aren't used on newer versions and the `GlStateManager` methods are even removed.

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/bridge/impl/AxoRenderContextImpl.java:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	@Override
+	public void drawTexture(int sx, int sy, int sw, int sh, int tx, int ty, int tw, int th, AxoIdentifier texture) {
```
It would probably preferable to keep the parameters in the same order as Minecraft in order to prevent bugs when porting since reordering `int`s does not induce errors.

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/bridge/impl/Bridge.java:
```diff
@@ -0,0 +1,45 @@
+
+	public static void init() {
+		KeyBindingEvents.REGISTER_KEYBINDS.register(keyBindingRegistry -> {
+			Preconditions.checkState(keyBindings != null, "double keybind register");
```
This condition is impossible to trigger as the `REGISTER_KEYBINDS` event is only called once (when the options file is loaded)

Floweynt replied:
This only exists for nullability analysis reasons, maybe `Objects.requireNonNull`?

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/bridge/mixin/internal/PlatformImplInternalMixin.java:
```diff
@@ -0,0 +1,127 @@
+	@Overwrite
+	public static AxoItemStack createItemStack(AxoItem item, int count) {
+		if (count == 0 || item == AxoItems.AIR) {
+			return new ItemStack(Item.byBlock(Blocks.STONE), 0);
```
What is the reason for this?

Floweynt replied:
Modern versions have support for "AIR" item, but 1.8.9 doesn't.

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/bridge/mixin/internal/PlatformImplInternalMixin.java:
```diff
@@ -0,0 +1,127 @@
+	 */
+	@Overwrite
+	public static AxoKeybinding createKeyBinding(AxoKey defaultKey, String name, String category) {
+		final var binding = new KeyBinding(name, ((AxoKeyImpl) defaultKey).id(), category);
```
You're also aware that LWJGL2 and GLFW use different key codes? Minecraft also saves keybindings differently across versions which will produce difficulties later.

Floweynt replied:
Yes, which is why `AxoKey` is designed as it is. To the common platform, it's a largely opaque "handle" for a key. For newer versions I might inject interface impl into  `InputUtil.Key` or whatever.

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/mixin/KeyBindingMixin.java:
```diff
@@ -48,7 +48,8 @@ public abstract class KeyBindingMixin {
 		if (pressed) {
-			Events.KEY_PRESS.invoker().invoke(new KeyPressEvent(key));
+			Events.KEY_INPUT.invoker().accept(key.getBoundKey());
+			io.github.axolotlclient.util.events.Events.KEY_PRESS.invoker().invoke(new KeyPressEvent(key));
```
What is the reason for this change?

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/modules/blur/MenuBlur.java:
```diff
@@ -22,6 +22,7 @@
 
 package io.github.axolotlclient.modules.blur;
 
+import io.github.axolotlclient.modules.AbstractModule;
```
This seems like an unnecessary change

moehreag commented on 1.8.9/src/main/resources/axolotlclient-bridge.mixins.json:
```diff
@@ -0,0 +1,31 @@
+	"injectors": {
+		"defaultRequire": 1
+	},
+	"mixins": [
```
These should probably also all go in the client array, even if it doesn't matter *that* much in the end since the mod just isn't going to get loaded on anything that isn't a client

moehreag commented on common/src/main/java/io/github/axolotlclient/AxolotlClientCommon.java:
```diff
@@ -1,77 +1,154 @@
-/*
```
Why did the header get removed?

moehreag commented on common/src/main/java/io/github/axolotlclient/AxolotlClientCommon.java:
```diff
@@ -1,77 +1,154 @@
-
 package io.github.axolotlclient;
 
-import java.time.format.DateTimeFormatter;
```
Ideally the imports should stay in the same order

moehreag commented on common/src/main/java/io/github/axolotlclient/AxolotlClientConfigCommon.java:
```diff
@@ -0,0 +1,29 @@
+	public final StringOption datetimeFormat = new StringOption("datetime_format", "yyyy/MM/dd HH:mm:ss", s -> dateTimeFormatter = DateTimeFormatter.ofPattern(s));
+
+	@Getter
+	private DateTimeFormatter dateTimeFormatter;
```
This field will not be initialized automatically by the option, it needs to be set once after the config is loaded.

moehreag commented on common/src/main/java/io/github/axolotlclient/bridge/Platform.java:
```diff
@@ -0,0 +1,24 @@
+		return PlatformImplInternal.getConfig();
+	}
+
+	static int getCurrentFps() {
```
This method should probably be moved to the client abstraction

moehreag commented on common/src/main/java/io/github/axolotlclient/bridge/events/EventBus.java:
```diff
@@ -0,0 +1,234 @@
+ *
+ * @param <T>
+ */
+public final class EventBus<T> {
```
It should be possible to simplify this class a lot as most features aren't used. I'm also not sure whether it's even necessary to begin with.

moehreag commented on common/src/main/java/io/github/axolotlclient/bridge/internal/BridgeUtil.java:
```diff
@@ -0,0 +1,23 @@
+	/**
+	 * Don't invoke any methods relating to {@link PlatformImplInternal}, so we need to get the version via {@link FabricLoader}
+	 */
+	private static final String MC_VERSION = FabricLoader.getInstance()
```
Can be replaced with a call to `AxolotlClientCommon#GAME_VERSION`

moehreag commented on common/src/main/java/io/github/axolotlclient/bridge/internal/PlatformImplInternal.java:
```diff
@@ -0,0 +1,71 @@
+		throw BridgeUtil.noImpl();
+	}
+
+	public static String getTranslatedString(String nameKey, Object[] args) {
```
There will also need to be accompanying methods for `Component`s/`Text`s

moehreag commented on common/src/main/java/io/github/axolotlclient/bridge/render/AxoRenderContext.java:
```diff
@@ -0,0 +1,137 @@
+		throw BridgeUtil.noImpl();
+	}
+
+	default void renderGuiItemOverlay(AxoItemStack stack, int x, int y, String countLabel, int textColor, boolean shadow) {
```
As parameters for text color and shadow are not available on newer versions it would make sense to fully hide them in the implementation for 1.8.9 instead of having to ignore them on newer versions.

moehreag commented on common/src/main/java/io/github/axolotlclient/bridge/render/AxoRenderContext.java:
```diff
@@ -0,0 +1,137 @@
+import io.github.axolotlclient.modules.hud.util.Rectangle;
+import org.jetbrains.annotations.ApiStatus;
+
+public interface AxoRenderContext {
```
Methods for scissoring should be added to this class as well

moehreag commented on common/src/main/java/io/github/axolotlclient/bridge/render/AxoTextRenderer.java:
```diff
@@ -0,0 +1,26 @@
+
+	default void drawCenteredString(AxoRenderContext context, String value, int x, int y, int color, boolean shadow) {
+		throw BridgeUtil.noImpl();
+	}
```
In my opinion these should only be exposed through the rendering abstraction, otherwise implementations on newer versions will be prone to bugs and infinite recursion.

moehreag commented on common/src/main/java/io/github/axolotlclient/bridge/world/AxoWorld.java:
```diff
@@ -0,0 +1,10 @@
+import io.github.axolotlclient.bridge.entity.AxoEntity;
+import java.util.List;
+
+public interface AxoWorld {
```
Topic: Level <-> World discussion

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hud/HudManager0.java:
```diff
@@ -0,0 +1,262 @@
+			}
+		} catch (IOException e) {
+			// TODO: better logging mechanism
+			// AxolotlClient.LOGGER.warn("Failed to load custom hud modules!", e);
```
The logger is already accessible through `AxolotlClientCommon#getLogger`

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hud/gui0/hud/MouseMovementHud.java:
```diff
@@ -0,0 +1,143 @@
+		context.drawTexture(spaceX + (width / 2) - 7 / 2 - 1, spaceY + 17 - (7 / 2), 0, 0, 7, 7, 7, 7, Platform.getTexture(mouseMovementIndicatorInner));
+		// Woah KodeToad, good use of translate
+		context.translateMatrix(calculatedMouseX, calculatedMouseY, 0);
+		context.drawTexture(spaceX + (width / 2) - 1, spaceY + 17, 0, 0, 11, 11, 11, 11, Platform.getTexture(mouseMovementIndicatorOuter));
```
I didn't even read this before writing the other comment about it, but here's the first bug related to the parameter order.

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hud/gui0/hud/item/ArmorHud.java:
```diff
@@ -0,0 +1,190 @@
+		}
+		String text = getItemString(stack);
+		int textY = y + 10 - context.getTextRenderer().getFontHeight() / 2;
+		float f = (float) stack.getDamage();
+		float g = (float) stack.getMaxDamage();
+		float h = Math.max(0.0F, (g - f) / g);
+		int j = java.awt.Color.HSBtoRGB(h / 3.0F, 1.0F, 1.0F);
```
This branch is out of date and this should also get moved to the implementation for 1.8.9 and 1.16, newer versions have `ItemStack#getBarColor` which should be used instead.

Floweynt replied:
Interesting, probably worth raising this to bridge then.

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hud/gui0/hud/item/ArrowHud.java:
```diff
@@ -20,70 +20,71 @@
-
-	public static final Identifier ID = new Identifier("kronhud", "arrowhud");
+	private static final AxoIdentifier ID = AxoIdentifier.of("kronhud", "arrowhud");
+	private static final AxoItemStack ITEM = AxoItemStack.of(AxoItems.ARROW);
```
This changes the modules function for newer versions. It should still also keep track of spectral & tipped arrows even if they do not exist on 1.8.9.

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hud/gui0/hud/simple/CPSHud.java:
```diff
@@ -42,29 +42,29 @@
 	private final BooleanOption fromKeybindings = new BooleanOption("cpskeybind", false);
 	private final BooleanOption rmb = new BooleanOption("rightcps", false);
 
+	private final ClickList left = new ClickList();
+	private final ClickList right = new ClickList();
```
These need to be accessible from the KeystrokesHud

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hud/gui0/hud/simple/CPSHud.java:
```diff
@@ -42,29 +42,29 @@
+					right.click();
+				}
+			} else {
+				if (key.equals(AxoKeys.MOUSE_LEFT)) {
```
As the keys' binding can be changed this is limiting this module's functionality.

Floweynt replied:
This just preserves the original behavior, which is keybind vs raw click type.

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hud/gui0/hud/simple/PingHud.java:
```diff
@@ -75,12 +60,12 @@ public boolean tickable() {
 			second++;
 	}
-
+/*
```
This should probably be implemented by just delegating to a method implemented in the bridge, since this ping code differs slightly across versions it really doesn't work in the common module.

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hud/gui0/hud/simple/ReachHud.java:
```diff
@@ -20,62 +20,60 @@
-		return camera.distanceTo(receiving.getRotationVec(1));
+	private static double getAttackDistance(AxoEntity attacking, AxoEntity receiving) {
+		Vec3 camera = attacking.getRotation(1);
+		return camera.dist(receiving.getRotation(1));
```
Other versions do a manual raycast here which *probably* works better. However, the method used is not available on 1.8.9.

moehreag reviewed (changes requested):
A lot of good things, but also still some major issues. Of course, this is nowhere near finished, but the more bugs can be fixed already the less we have to find afterwards.

Commit:
cleanup based on github comments

Floweynt commented:
Resolved some of the comments. The code currently is ugly and has extraneous diffs, but I will clean them up later with some automated tooling. I just want to get everything in a working state before I deal with that.

Commit:
fix compilation on some other versions (1.21.4 is not implemented because mojmap)

note that this commit DOES NOT implement proper functionality (events, calling the correct render events, etc).

Commit:
preliminary bridge implementation for 1.21.4

Commit:
begin porting hypixel modules

Commit:
scissor and some other stuff

Commit:
Things compile now

Commit:
Get rid of hypixel modules

Commit:
Fix some startup crashes + things not building for some reason.

Floweynt commented:
Limiting the scope of this now. Only porting Hypixel and HUD modules, and not even all HUD modules. I want this to be in an usable state in a few months, and being testing soon.

Future work can include porting other modules / cleanup.

Commit:
migrate 1.8.9 to the common hud system rather than hosting both systems.

Commit:
Improve bridge implementation on 1.8.9

Commit:
squashme

Commit:
cleanup for 1.16

Commit:
hud now doesn't crash

Commit:
1.20 runs without crashing, mostly

Commit:
add license headers

Commit:
"stable" for 1.21

Commit:
merge "properly"

Commit:
implement new bridge apis for 1.8.9

Commit:
merge dev, again

Commit:
implement new bridge api for 1.16

Commit:
bridge clean up & commands

Commit:
merge dev

Commit:
bug fixes

Commit:
uncomment some TODOs

Commit:
fix config file logic

Commit:
validate bridge implementations for 1.8.9, 1.16, and 1.20

Commit:
validate most (all?) bridge implementations

Commit:
license header fixes

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/HypixelMessages.java:
```diff
@@ -51,7 +45,8 @@ public void load() {
 
-		AxolotlClient.LOGGER.debug("Loading Hypixel Messages");
+		AxolotlClientCommon.getInstance().getLogger().debug("Loading Hypixel Messages");
+		/*
```
should probably add a `TODO` or something to not forget about this

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/AutoTip.java:
```diff
@@ -65,13 +61,14 @@ public OptionCategory getCategory() {
 	@Override
 	public void tick() {
 		if (init) {
+			/*
```
This could probably be implemented now. There's also the potential for improvement on the server check but this would be a bonus point.

moehreag commented on common/src/main/java/io/github/axolotlclient/bridge/util/AxoText.java:
```diff
@@ -0,0 +1,153 @@
+
+	interface Style {
+		@RequiresImpl
+		default Style br$color(Color color) {
```
There should probably also be a convenience method for the `Color` class in AxolotlClientConfig (adapting using `.toInt()`)

moehreag commented on 1.8.9/src/main/resources/axolotlclient-bridge.mixins.json:
```diff
@@ -0,0 +1,40 @@
+        "defaultRequire": 1
+    },
+    "mixins": [
+        "util.StyleMixin"
```
Why isn't this mixin in the client block as well? It doesn't make a difference as the environment is specified to be client-only in the fmj anyways.

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/mixin/LivingEntityMixin.java:
```diff
@@ -69,12 +66,23 @@ public LivingEntityMixin(World world) {
+			ComboHud comboHud = (ComboHud) HudManager.getInstance().get(ComboHud.ID);
+			comboHud.onEntityDamage(this);
+		}
+=======
```
this seems to be a remainder of a conflict

moehreag commented on 1.21.7/src/main/java/io/github/axolotlclient/bridge/mixin/render/FontMixin.java:
```diff
@@ -0,0 +1,60 @@
+
+	@Override
+	public int br$getWidth(AxoText text) {
+		return this.br$getWidth(((Component) text).getString());
```
This should probably directly use minecraft's method without unwrapping the component first (without calling `getString`)

moehreag commented on 1.16_combat-6/src/main/java/io/github/axolotlclient/modules/hud/gui/hud/KeystrokeHud.java:
```diff
@@ -155,25 +163,25 @@ public Keystroke createFromString(Rectangle bounds, DrawPosition offset, KeyBind
 		}
 		for (Keystroke stroke : keystrokes) {
-			stroke.render(graphics);
+			stroke.render((MatrixStack) graphics);
```
I'm not too happy about these casts, ideally it should use the render context impl all the way down

moehreag reviewed (commented):
- I like a lot of things about this
- There are still some things to do left

We'd prefer to orient names on mojmap generally but that's a cosmetic issue which can also be changed/discussed afterwards.
Ideally it'd be nice to keep the import order as well.

Commit:
update config version manager & fix NPE with datetimeformatter

Floweynt committed:
resolve some comments

Floweynt committed:
fix mixin config

Floweynt committed:
fix imports

moehreag committed:
bump gradle & loom

improve user agent format

moehreag committed:
cleanup user agent

moehreag committed:
ensure config loading

Floweynt committed:
attempt some bugfixes

moehreag committed:
fix a lot of bugs

- too early initialization of hud bounds
- potion hud sprites
- wrong parameters in gui mixins
- icon hud getting cut off
- trap upgrade icons
- add a few todos

moehreag committed:
bump config lib

moehreag committed:
add keybind to toggle hud rendering

moehreag committed:
use `consumeClick` handlers for keybindings

moehreag committed:
add bridge API for resource manager access

moehreag committed:
Merge remote-tracking branch 'origin/dev' into flowey/bridge

# Conflicts:
#	1.16_combat-6/src/main/java/io/github/axolotlclient/AxolotlClient.java
#	1.16_combat-6/src/main/java/io/github/axolotlclient/modules/hud/HudManager.java
#	1.16_combat-6/src/main/java/io/github/axolotlclient/modules/hud/gui/hud/PotionsHud.java
#	1.16_combat-6/src/main/java/io/github/axolotlclient/modules/hypixel/StatsMod.java
#	1.20/src/main/java/io/github/axolotlclient/AxolotlClient.java
#	1.20/src/main/java/io/github/axolotlclient/modules/hud/HudManager.java
#	1.20/src/main/java/io/github/axolotlclient/modules/hud/gui/hud/PotionsHud.java
#	1.20/src/main/java/io/github/axolotlclient/modules/hypixel/StatsMod.java
#	1.21.7/build.gradle.kts
#	1.21.7/src/main/java/io/github/axolotlclient/AxolotlClient.java
#	1.21.7/src/main/java/io/github/axolotlclient/mixin/PlayerEntityRendererMixin.java
#	1.21.7/src/main/java/io/github/axolotlclient/modules/hud/HudManager.java
#	1.21.7/src/main/java/io/github/axolotlclient/modules/hud/gui/hud/PotionsHud.java
#	1.21.7/src/main/java/io/github/axolotlclient/modules/hypixel/StatsMod.java
#	1.21/src/main/java/io/github/axolotlclient/AxolotlClient.java
#	1.21/src/main/java/io/github/axolotlclient/modules/hud/HudManager.java
#	1.21/src/main/java/io/github/axolotlclient/modules/hud/gui/hud/PotionsHud.java
#	1.21/src/main/java/io/github/axolotlclient/modules/hypixel/StatsMod.java
#	1.8.9/src/main/java/io/github/axolotlclient/AxolotlClient.java
#	1.8.9/src/main/java/io/github/axolotlclient/modules/hud/HudManager.java
#	1.8.9/src/main/java/io/github/axolotlclient/modules/hud/gui/hud/PotionsHud.java
#	1.8.9/src/main/java/io/github/axolotlclient/modules/hypixel/StatsMod.java
#	common/src/main/java/io/github/axolotlclient/AxolotlClientCommon.java
#	common/src/main/resources/assets/axolotlclient/lang/en_us.json
#	gradle.properties

moehreag committed:
fix another bug

moehreag committed:
disable mc jar verification for 1.16_combat-6

moehreag committed:
the same but so that it hopefully works

moehreag committed:
clean up main class

moehreag committed:
Implement bridge for AutoBoop filters

- store filters as base64 to prevent conflicts with commas

moehreag committed:
bump config lib + fabric loader version

moehreag committed:
fix resource reloads

moehreag committed:
add missing AxoSprites Mixin to 1.21.8

moehreag committed:
(1.8.9) fix hotbar hud

moehreag committed:
Implement `PlayerHud`s

moehreag committed:
rename common hud package

moehreag committed:
clean up a bit
